File: Heap Priority Queue/0253. Meeting Rooms II/0253.py
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        intervals.sort(key=lambda x:(x[0]))
        min_heap = [intervals[0][1]]

        for i in range(1, len(intervals)):
            if intervals[i][0] >= min_heap[0]:
                heapq.heappop(min_heap)
            heapq.heappush(min_heap, intervals[i][1])
            
        return len(min_heap)

File: Heap Priority Queue/0253. Meeting Rooms II/0253.cpp
class Solution {
public:
    int minMeetingRooms(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), [](const vector<int>&a, const vector<int>&b)
        {
            return a[0] < b[0];
        });

        priority_queue<int, vector<int>, greater<int>> pq;
        pq.push(intervals[0][1]);

        for (int i=0; i<intervals.size(); ++i) {
            if (intervals[i][0] >= pq.top()) {
                pq.pop();
            } pq.push(intervals[i][1]);
        } return pq.size();
    }
};